sklearn/metrics/cluster/_unsupervised.py

```diff
@@ -131,15 +132,27 @@ def _silhouette_reduce(D_chunk, start, labels, label_freqs):
     label_freqs : array-like
         Distribution of cluster labels in ``labels``.
     """
+    n_chunk_samples = D_chunk.shape[0]
     # accumulate distances from each sample to each cluster
-    clust_dists = np.zeros((len(D_chunk), len(label_freqs)), dtype=D_chunk.dtype)
-    for i in range(len(D_chunk)):
+    clust_dists = np.zeros((n_chunk_samples, len(label_freqs)), dtype=D_chunk.dtype)
+    D_chunk_is_sparse = issparse(D_chunk)
+
+    for i in range(n_chunk_samples):
+        if D_chunk_is_sparse:
```

Comment:
I don't really like the fact that we do a check for sparsity within the `for` loop.
I would inverse both:

```python
if issparse(D_chunk):
    for ...
else:
    for ...
```

Reply:
also I would add a check that the matrix is for CSR format:

```python
if issparse(D_chunck):
    if D_chunk.format != "csr":
        raise TypeError(...)
    for ...
...
```